File: llama_mobile_vd-android-java-SDK/src/main/java/com/llamamobile/vd/LlamaMobileVD.java
package com.llamamobile.vd;

public class LlamaMobileVD {
    static {
        System.loadLibrary("llama_mobile_vd_jni");
    }
    
    // Distance Metric enum
    public enum DistanceMetric {
        L2(0),
        COSINE(1),
        DOT(2);
        
        private final int value;
        
        DistanceMetric(int value) {
            this.value = value;
        }
        
        public int getValue() {
            return value;
        }
    }
    
    // VectorStore native methods
    public static native long nativeVectorStoreCreate(int dimension, int metric);
    public static native void nativeVectorStoreAddVector(long storeId, long id, float[] vector);
    public static native SearchResult[] nativeVectorStoreSearch(long storeId, float[] queryVector, int k);
    public static native float[] nativeVectorStoreGetVector(long storeId, long id);
    public static native boolean nativeVectorStoreRemoveVector(long storeId, long id);
    public static native boolean nativeVectorStoreContains(long storeId, long id);
    public static native long nativeVectorStoreGetSize(long storeId);
    public static native int nativeVectorStoreGetDimension(long storeId);
    public static native int nativeVectorStoreGetMetric(long storeId);
    public static native boolean nativeVectorStoreUpdateVector(long storeId, long id, float[] vector);
    public static native boolean nativeVectorStoreReserve(long storeId, long capacity);
    public static native void nativeVectorStoreClear(long storeId);
    public static native void nativeVectorStoreDestroy(long storeId);
    
    // HNSWIndex native methods
    public static native long nativeHNSWIndexCreate(int dimension, int metric, long maxElements);
    public static native long nativeHNSWIndexCreateWithParams(int dimension, int metric, long maxElements, int M, int efConstruction, int seed);
    public static native boolean nativeHNSWIndexAddVector(long indexId, long id, float[] vector);
    public static native SearchResult[] nativeHNSWIndexSearch(long indexId, float[] queryVector, int k);
    public static native boolean nativeHNSWIndexSetEfSearch(long indexId, int efSearch);
    public static native int nativeHNSWIndexGetEfSearch(long indexId);
    public static native long nativeHNSWIndexGetSize(long indexId);
    public static native int nativeHNSWIndexGetDimension(long indexId);
    public static native long nativeHNSWIndexGetCapacity(long indexId);
    public static native boolean nativeHNSWIndexContains(long indexId, long id);
    public static native float[] nativeHNSWIndexGetVector(long indexId, long id);
    public static native boolean nativeHNSWIndexSave(long indexId, String filename);
    public static native long nativeHNSWIndexLoad(String filename);
    public static native void nativeHNSWIndexDestroy(long indexId);
    
    // MMapVectorStoreBuilder native methods
    public static native long nativeMMapVectorStoreBuilderCreate(int dimension, int metric);
    public static native boolean nativeMMapVectorStoreBuilderAddVector(long builderId, long id, float[] vector);
    public static native boolean nativeMMapVectorStoreBuilderReserve(long builderId, long capacity);
    public static native boolean nativeMMapVectorStoreBuilderSave(long builderId, String filename);
    public static native long nativeMMapVectorStoreBuilderGetSize(long builderId);
    public static native int nativeMMapVectorStoreBuilderGetDimension(long builderId);
    public static native void nativeMMapVectorStoreBuilderDestroy(long builderId);
    
    // MMapVectorStore native methods
    public static native long nativeMMapVectorStoreOpen(String filename);
    public static native float[] nativeMMapVectorStoreGetVector(long storeId, long id);
    public static native boolean nativeMMapVectorStoreContains(long storeId, long id);
    public static native SearchResult[] nativeMMapVectorStoreSearch(long storeId, float[] queryVector, int k);
    public static native long nativeMMapVectorStoreGetSize(long storeId);
    public static native int nativeMMapVectorStoreGetDimension(long storeId);
    public static native int nativeMMapVectorStoreGetMetric(long storeId);
    public static native void nativeMMapVectorStoreClose(long storeId);
    
    // Version information native methods
    public static native String nativeGetVersion();
    public static native int nativeGetVersionMajor();
    public static native int nativeGetVersionMinor();
    public static native int nativeGetVersionPatch();
    
    // Convenience methods for VectorStore
    public static long createVectorStore(int dimension, DistanceMetric metric) {
        return nativeVectorStoreCreate(dimension, metric.getValue());
    }
    
    public static long createVectorStore(int dimension) {
        return createVectorStore(dimension, DistanceMetric.COSINE);
    }
    
    // Convenience methods for HNSWIndex
    public static long createHNSWIndex(int dimension, DistanceMetric metric, long maxElements) {
        return nativeHNSWIndexCreate(dimension, metric.getValue(), maxElements);
    }
    
    public static long createHNSWIndex(int dimension, DistanceMetric metric, long maxElements, int M, int efConstruction, int seed) {
        return nativeHNSWIndexCreateWithParams(dimension, metric.getValue(), maxElements, M, efConstruction, seed);
    }
    
    public static long createHNSWIndex(int dimension, long maxElements) {
        return createHNSWIndex(dimension, DistanceMetric.COSINE, maxElements);
    }
    
    // Convenience methods for MMapVectorStoreBuilder
    public static long createMMapVectorStoreBuilder(int dimension, DistanceMetric metric) {
        return nativeMMapVectorStoreBuilderCreate(dimension, metric.getValue());
    }
    
    public static long createMMapVectorStoreBuilder(int dimension) {
        return createMMapVectorStoreBuilder(dimension, DistanceMetric.COSINE);
    }
    
    public static void saveMMapVectorStoreBuilder(long builderId, String filename) {
        nativeMMapVectorStoreBuilderSave(builderId, filename);
    }
    
    // Convenience methods for MMapVectorStore
    public static long openMMapVectorStore(String filePath) {
        return nativeMMapVectorStoreOpen(filePath);
    }
    
    // Version information methods
    public static String getVersion() {
        return nativeGetVersion();
    }
    
    public static int getVersionMajor() {
        return nativeGetVersionMajor();
    }
    
    public static int getVersionMinor() {
        return nativeGetVersionMinor();
    }
    
    public static int getVersionPatch() {
        return nativeGetVersionPatch();
    }
    
    // SearchResult class for search query results
    public static class SearchResult {
        private final long id;
        private final float distance;
        
        public SearchResult(long id, float distance) {
            this.id = id;
            this.distance = distance;
        }
        
        public long getId() {
            return id;
        }
        
        public float getDistance() {
            return distance;
        }
        
        @Override
        public String toString() {
            return "SearchResult{id=" + id + ", distance=" + distance + "}";
        }
    }
    
    // LlamaMobileVDException class for error handling
    public static class LlamaMobileVDException extends RuntimeException {
        public LlamaMobileVDException(String message) {
            super(message);
        }
        
        public LlamaMobileVDException(String message, Throwable cause) {
            super(message, cause);
        }
    }
}

File: examples/androidSDKExample/app/src/main/java/com/llamamobile/vd/LlamaMobileVD.kt
package com.llamamobile.vd

/**
 * LlamaMobileVD Android SDK - Kotlin Wrapper
 * A high-performance vector database for Android applications.
 */
class LlamaMobileVD {

    companion object {
        init {
            System.loadLibrary("llama_mobile_vd_jni")
        }

        /**
         * Gets the version of the LlamaMobileVD library.
         */
        @JvmStatic
        external fun getVersion(): String
    }

    /**
     * Distance metrics supported by LlamaMobileVD.
     */
    enum class DistanceMetric {
        L2,
        COSINE,
        DOT
    }

    /**
     * Search result containing ID and distance.
     */
    data class SearchResult(val id: Long, val distance: Float)

    /**
     * VectorStore interface for managing vectors in memory.
     */
    class VectorStore(
        dimension: Int,
        metric: DistanceMetric
    ) : AutoCloseable {
        private val handle: Long

        init {
            handle = nativeVectorStoreCreate(dimension, metric.ordinal)
        }

        /**
         * Adds a vector to the store with the specified ID.
         */
        fun add(id: Long, vector: FloatArray) {
            nativeVectorStoreAdd(handle, id, vector)
        }

        /**
         * Removes a vector from the store with the specified ID.
         * @return true if the vector was removed, false otherwise
         */
        fun remove(id: Long): Boolean {
            return nativeVectorStoreRemove(handle, id)
        }

        /**
         * Gets a vector from the store with the specified ID.
         */
        fun get(id: Long): FloatArray {
            return nativeVectorStoreGet(handle, id)
        }

        /**
         * Updates a vector in the store with the specified ID.
         */
        fun update(id: Long, vector: FloatArray) {
            nativeVectorStoreUpdate(handle, id, vector)
        }

        /**
         * Searches for the k nearest neighbors to the query vector.
         */
        fun search(query: FloatArray, k: Int): Array<SearchResult> {
            return nativeVectorStoreSearch(handle, query, k)
        }

        /**
         * Gets the number of vectors in the store.
         */
        fun size(): Int {
            return nativeVectorStoreSize(handle)
        }

        /**
         * Gets the dimension of vectors in the store.
         */
        fun dimension(): Int {
            return nativeVectorStoreDimension(handle)
        }

        /**
         * Gets the distance metric used by the store.
         */
        fun metric(): DistanceMetric {
            return DistanceMetric.values()[nativeVectorStoreMetric(handle)]
        }

        /**
         * Checks if the store contains a vector with the specified ID.
         */
        fun contains(id: Long): Boolean {
            return nativeVectorStoreContains(handle, id)
        }

        /**
         * Reserves capacity for the specified number of vectors.
         */
        fun reserve(capacity: Int) {
            nativeVectorStoreReserve(handle, capacity)
        }

        /**
         * Clears all vectors from the store.
         */
        fun clear() {
            nativeVectorStoreClear(handle)
        }

        /**
         * Releases the native resources.
         */
        override fun close() {
            nativeVectorStoreDestroy(handle)
        }

        private external fun nativeVectorStoreCreate(dimension: Int, metric: Int): Long
        private external fun nativeVectorStoreAdd(handle: Long, id: Long, vector: FloatArray)
        private external fun nativeVectorStoreRemove(handle: Long, id: Long): Boolean
        private external fun nativeVectorStoreGet(handle: Long, id: Long): FloatArray
        private external fun nativeVectorStoreUpdate(handle: Long, id: Long, vector: FloatArray)
        private external fun nativeVectorStoreSearch(handle: Long, query: FloatArray, k: Int): Array<SearchResult>
        private external fun nativeVectorStoreSize(handle: Long): Int
        private external fun nativeVectorStoreDimension(handle: Long): Int
        private external fun nativeVectorStoreMetric(handle: Long): Int
        private external fun nativeVectorStoreContains(handle: Long, id: Long): Boolean
        private external fun nativeVectorStoreReserve(handle: Long, capacity: Int)
        private external fun nativeVectorStoreClear(handle: Long)
        private external fun nativeVectorStoreDestroy(handle: Long)
    }

    /**
     * HNSWIndex interface for efficient approximate nearest neighbor search.
     */
    class HNSWIndex private constructor(private val handle: Long) : AutoCloseable {
        /**
         * Creates a new HNSWIndex with default parameters.
         */
        constructor(
            dimension: Int,
            metric: DistanceMetric,
            maxElements: Int
        ) : this(Companion.nativeHNSWIndexCreate(dimension, metric.ordinal, maxElements))

        /**
         * Creates a new HNSWIndex with custom parameters.
         */
        constructor(
            dimension: Int,
            metric: DistanceMetric,
            maxElements: Int,
            M: Int,
            efConstruction: Int,
            seed: Int
        ) : this(Companion.nativeHNSWIndexCreateWithParams(dimension, metric.ordinal, maxElements, M, efConstruction, seed))

        /**
         * Adds a vector to the index with the specified ID.
         */
        fun add(id: Long, vector: FloatArray) {
            Companion.nativeHNSWIndexAdd(handle, id, vector)
        }

        /**
         * Searches for the k nearest neighbors to the query vector.
         */
        fun search(query: FloatArray, k: Int): Array<SearchResult> {
            return Companion.nativeHNSWIndexSearch(handle, query, k)
        }

        /**
         * Sets the ef_search parameter for the index.
         */
        fun setEfSearch(efSearch: Int) {
            Companion.nativeHNSWIndexSetEfSearch(handle, efSearch)
        }

        /**
         * Gets the current ef_search parameter value.
         */
        fun getEfSearch(): Int {
            return Companion.nativeHNSWIndexGetEfSearch(handle)
        }

        /**
         * Gets the number of vectors in the index.
         */
        fun size(): Int {
            return Companion.nativeHNSWIndexSize(handle)
        }

        /**
         * Gets the dimension of vectors in the index.
         */
        fun dimension(): Int {
            return Companion.nativeHNSWIndexDimension(handle)
        }

        /**
         * Gets the capacity of the index.
         */
        fun capacity(): Int {
            return Companion.nativeHNSWIndexCapacity(handle)
        }

        /**
         * Checks if the index contains a vector with the specified ID.
         */
        fun contains(id: Long): Boolean {
            return Companion.nativeHNSWIndexContains(handle, id)
        }

        /**
         * Gets a vector from the index with the specified ID.
         */
        fun getVector(id: Long): FloatArray {
            return Companion.nativeHNSWIndexGetVector(handle, id)
        }

        /**
         * Saves the index to a file.
         */
        fun save(filename: String) {
            Companion.nativeHNSWIndexSave(handle, filename)
        }

        /**
         * Releases the native resources.
         */
        override fun close() {
            Companion.nativeHNSWIndexDestroy(handle)
        }

        companion object {
            /**
             * Loads an HNSWIndex from a file.
             */
            fun load(filename: String): HNSWIndex {
                val handle = nativeHNSWIndexLoad(filename)
                return HNSWIndex(handle)
            }

            private external fun nativeHNSWIndexCreate(dimension: Int, metric: Int, maxElements: Int): Long
            private external fun nativeHNSWIndexCreateWithParams(dimension: Int, metric: Int, maxElements: Int, M: Int, efConstruction: Int, seed: Int): Long
            private external fun nativeHNSWIndexAdd(handle: Long, id: Long, vector: FloatArray)
            private external fun nativeHNSWIndexSearch(handle: Long, query: FloatArray, k: Int): Array<SearchResult>
            private external fun nativeHNSWIndexSetEfSearch(handle: Long, efSearch: Int)
            private external fun nativeHNSWIndexGetEfSearch(handle: Long): Int
            private external fun nativeHNSWIndexSize(handle: Long): Int
            private external fun nativeHNSWIndexDimension(handle: Long): Int
            private external fun nativeHNSWIndexCapacity(handle: Long): Int
            private external fun nativeHNSWIndexContains(handle: Long, id: Long): Boolean
            private external fun nativeHNSWIndexGetVector(handle: Long, id: Long): FloatArray
            private external fun nativeHNSWIndexSave(handle: Long, filename: String)
            private external fun nativeHNSWIndexLoad(filename: String): Long
            private external fun nativeHNSWIndexDestroy(handle: Long)
        }
    }

    /**
     * MMapVectorStoreBuilder for creating memory-mapped vector stores on disk.
     */
    class MMapVectorStoreBuilder(dimension: Int, metric: DistanceMetric) : AutoCloseable {
        private val handle: Long

        init {
            handle = nativeMMapVectorStoreBuilderCreate(dimension, metric.ordinal)
        }

        /**
         * Adds a vector to the builder.
         */
        fun add(id: Long, vector: FloatArray) {
            nativeMMapVectorStoreBuilderAdd(handle, id, vector)
        }

        /**
         * Reserves capacity for the specified number of vectors.
         */
        fun reserve(capacity: Int) {
            nativeMMapVectorStoreBuilderReserve(handle, capacity)
        }

        /**
         * Saves the builder to a file.
         */
        fun save(filename: String) {
            nativeMMapVectorStoreBuilderSave(handle, filename)
        }

        /**
         * Gets the number of vectors in the builder.
         */
        fun size(): Int {
            return nativeMMapVectorStoreBuilderSize(handle)
        }

        /**
         * Gets the dimension of vectors in the builder.
         */
        fun dimension(): Int {
            return nativeMMapVectorStoreBuilderDimension(handle)
        }

        override fun close() {
            nativeMMapVectorStoreBuilderDestroy(handle)
        }

        private external fun nativeMMapVectorStoreBuilderCreate(dimension: Int, metric: Int): Long
        private external fun nativeMMapVectorStoreBuilderAdd(handle: Long, id: Long, vector: FloatArray)
        private external fun nativeMMapVectorStoreBuilderReserve(handle: Long, capacity: Int)
        private external fun nativeMMapVectorStoreBuilderSave(handle: Long, filename: String)
        private external fun nativeMMapVectorStoreBuilderSize(handle: Long): Int
        private external fun nativeMMapVectorStoreBuilderDimension(handle: Long): Int
        private external fun nativeMMapVectorStoreBuilderDestroy(handle: Long)
    }

    /**
     * MMapVectorStore for accessing memory-mapped vector stores on disk.
     */
    class MMapVectorStore(filename: String) : AutoCloseable {
        private val handle: Long

        init {
            handle = nativeMMapVectorStoreOpen(filename)
        }

        /**
         * Gets a vector by its ID.
         */
        fun get(id: Long): FloatArray {
            return nativeMMapVectorStoreGet(handle, id)
        }

        /**
         * Checks if a vector with the specified ID exists.
         */
        fun contains(id: Long): Boolean {
            return nativeMMapVectorStoreContains(handle, id)
        }

        /**
         * Searches for the k nearest neighbors to the query vector.
         */
        fun search(query: FloatArray, k: Int): Array<SearchResult> {
            return nativeMMapVectorStoreSearch(handle, query, k)
        }

        /**
         * Gets the number of vectors in the store.
         */
        fun size(): Int {
            return nativeMMapVectorStoreSize(handle)
        }

        /**
         * Gets the dimension of vectors in the store.
         */
        fun dimension(): Int {
            return nativeMMapVectorStoreDimension(handle)
        }

        /**
         * Gets the distance metric used by the store.
         */
        fun metric(): DistanceMetric {
            return DistanceMetric.values()[nativeMMapVectorStoreMetric(handle)]
        }

        override fun close() {
            nativeMMapVectorStoreClose(handle)
        }

        private external fun nativeMMapVectorStoreOpen(filename: String): Long
        private external fun nativeMMapVectorStoreGet(handle: Long, id: Long): FloatArray
        private external fun nativeMMapVectorStoreContains(handle: Long, id: Long): Boolean
        private external fun nativeMMapVectorStoreSearch(handle: Long, query: FloatArray, k: Int): Array<SearchResult>
        private external fun nativeMMapVectorStoreSize(handle: Long): Int
        private external fun nativeMMapVectorStoreDimension(handle: Long): Int
        private external fun nativeMMapVectorStoreMetric(handle: Long): Int
        private external fun nativeMMapVectorStoreClose(handle: Long)
    }

    /**
     * Exception thrown by LlamaMobileVD operations.
     */
    class LlamaMobileVDException(message: String) : Exception(message)

    // Native methods
    private external fun nativeGetVersion(): String
}